Thats super weird... is it due to differing instances types maybe? On my devbox it's failing on master without the pytest fix:
```
(myenv) ubuntu@devbox:~/ray$ python -m pytest python/ray/tests/test_actor_failures.py::test_exit_actor_queued -v -s
======================================================================================== test session starts =========================================================================================
platform linux -- Python 3.9.23, pytest-8.4.1, pluggy-1.6.0 -- /home/ubuntu/.conda/envs/myenv/bin/python
cachedir: .pytest_cache
rootdir: /home/ubuntu/ray
configfile: pytest.ini
plugins: repeat-0.9.4, asyncio-1.2.0, anyio-4.9.0, retry-1.7.0
asyncio: mode=strict, debug=False, asyncio_default_fixture_loop_scope=None, asyncio_default_test_loop_scope=function
collected 1 item                                                                                                                                                                                     

python/ray/tests/test_actor_failures.py::test_exit_actor_queued 2025-10-02 02:41:12,951	INFO worker.py:2004 -- Started a local Ray instance. View the dashboard at 127.0.0.1:8265 
(raylet) Warning: More than 5000 tasks are pending submission to actor eea8dc9a62b8cb8de3bdff5f01000000. To reduce memory usage, wait for these tasks to finish before sending more.
FAILED(raylet) Warning: More than 10000 tasks are pending submission to actor c87d9a0c6623219fd5c695fc01000000. To reduce memory usage, wait for these tasks to finish before sending more.


============================================================================================== FAILURES ==============================================================================================
_______________________________________________________________________________________ test_exit_actor_queued _______________________________________________________________________________________

shutdown_only = None

    def test_exit_actor_queued(shutdown_only):
        """Verify after exit_actor is called the queued tasks won't execute."""
    
        @ray.remote
        class RegressionSync:
            def f(self):
                import time
    
                time.sleep(1)
                exit_actor()
    
            def ping(self):
                pass
    
        @ray.remote
        class RegressionAsync:
            async def f(self):
                await asyncio.sleep(1)
                exit_actor()
    
            def ping(self):
                pass
    
        # Test async case.
        # https://github.com/ray-project/ray/issues/32376
        # If we didn't fix this issue, this will segfault.
        a = RegressionAsync.remote()
        a.f.remote()
        refs = [a.ping.remote() for _ in range(10000)]
        with pytest.raises(ray.exceptions.RayActorError
        ) as exc_info:
            ray.get(refs)
        assert " Worker unexpectedly exits" not in str(exc_info.value)
    
        # Test a sync case.
        a = RegressionSync.remote()
        a.f.remote()
        with pytest.raises(ray.exceptions.RayActorError
        ) as exc_info:
>           ray.get([a.ping.remote() for _ in range(10000)])

python/ray/tests/test_actor_failures.py:1196: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
python/ray/_private/auto_init_hook.py:22: in auto_init_wrapper
    return fn(*args, **kwargs)
python/ray/_private/client_mode_hook.py:104: in wrapper
    return func(*args, **kwargs)
python/ray/_private/worker.py:2962: in get
    values, debugger_breakpoint = worker.get_objects(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <ray._private.worker.Worker object at 0x7f77e2a43730>
object_refs = [ObjectRef(9078e25b3cd9207ac87d9a0c6623219fd5c695fc0100000001000000), ObjectRef(f8570aca6e94fb7ac87d9a0c6623219fd5c695...4f4c87d9a0c6623219fd5c695fc0100000001000000), ObjectRef(852ddd9a96c844dcc87d9a0c6623219fd5c695fc0100000001000000), ...]
timeout = None, return_exceptions = False, skip_deserialization = False, _tensor_transport = None

    def get_objects(
        self,
        object_refs: list,
        timeout: Optional[float] = None,
        return_exceptions: bool = False,
        skip_deserialization: bool = False,
        _tensor_transport: Optional[str] = None,
    ) -> Tuple[List[serialization.SerializedRayObject], bytes]:
        """Get the values in the object store associated with the IDs.
    
        Return the values from the local object store for object_refs. This
        will block until all the values for object_refs have been written to
        the local object store.
    
        Args:
            object_refs: A list of the object refs
                whose values should be retrieved.
            timeout: The maximum amount of time in
                seconds to wait before returning.
            return_exceptions: If any of the objects deserialize to an
                Exception object, whether to return them as values in the
                returned list. If False, then the first found exception will be
                raised.
            skip_deserialization: If true, only the buffer will be released and
                the object associated with the buffer will not be deserialized.
            _tensor_transport: [Alpha] The tensor transport to use to fetch `torch.Tensors` found in the Ray Direct Transport object. Currently, this supports "object_store" and "nixl".
        Returns:
            list: List of deserialized objects or None if skip_deserialization is True.
            bytes: UUID of the debugger breakpoint we should drop
                into or b"" if there is no breakpoint.
        """
        # Make sure that the values are object refs.
        for object_ref in object_refs:
            if not isinstance(object_ref, ObjectRef):
                raise TypeError(
                    f"Attempting to call `get` on the value {object_ref}, "
                    "which is not an ray.ObjectRef."
                )
        tensor_transport: TensorTransportEnum = (
            TensorTransportEnum.from_str(_tensor_transport)
            if _tensor_transport is not None
            else None
        )
        assert tensor_transport in [
            TensorTransportEnum.OBJECT_STORE,
            TensorTransportEnum.NIXL,
            None,
        ], "Currently, RDT only supports 'object_store' and 'nixl' for tensor transport in ray.get()."
        timeout_ms = (
            int(timeout * 1000) if timeout is not None and timeout != -1 else -1
        )
        serialized_objects: List[
            serialization.SerializedRayObject
        ] = self.core_worker.get_objects(
            object_refs,
            timeout_ms,
        )
    
        debugger_breakpoint = b""
        for data, metadata, _ in serialized_objects:
            if metadata:
                metadata_fields = metadata.split(b",")
                if len(metadata_fields) >= 2 and metadata_fields[1].startswith(
                    ray_constants.OBJECT_METADATA_DEBUG_PREFIX
                ):
                    debugger_breakpoint = metadata_fields[1][
                        len(ray_constants.OBJECT_METADATA_DEBUG_PREFIX) :
                    ]
        if skip_deserialization:
            return None, debugger_breakpoint
    
        values = self.deserialize_objects(
            serialized_objects, object_refs, tensor_transport_hint=tensor_transport
        )
        if not return_exceptions:
            # Raise exceptions instead of returning them to the user.
            for i, value in enumerate(values):
                if isinstance(value, RayError):
                    if isinstance(value, ray.exceptions.ObjectLostError):
                        global_worker.core_worker.log_plasma_usage()
                    if isinstance(value, RayTaskError):
                        raise value.as_instanceof_cause()
                    else:
>                       raise value
E                       ray.exceptions.TaskCancelledError: The task aa9c16e4326c9c3ec87d9a0c6623219fd5c695fc01000000 is canceled from an actor c87d9a0c6623219fd5c695fc01000000 before it executes.

python/ray/_private/worker.py:1028: TaskCancelledError
====================================================================================== short test summary info =======================================================================================
FAILED python/ray/tests/test_actor_failures.py::test_exit_actor_queued - ray.exceptions.TaskCancelledError: The task aa9c16e4326c9c3ec87d9a0c6623219fd5c695fc01000000 is canceled from an actor c87d9a0c6623219fd5c695fc01000000 before it executes.
```

And here is the automated issue that flagged some failing postmerge tests: https://github.com/ray-project/ray/issues/44058#issuecomment-3354759849

--- a/python/ray/tests/test_actor_failures.py
+++ b/python/ray/tests/test_actor_failures.py
@@ -1183,18 +1183,14 @@ def test_exit_actor_queued(shutdown_only):
     a = RegressionAsync.remote()
     a.f.remote()
     refs = [a.ping.remote() for _ in range(10000)]
-    with pytest.raises(
-        (ray.exceptions.RayActorError, ray.exceptions.TaskCancelledError)
-    ) as exc_info:
+    with pytest.raises(ray.exceptions.RayActorError) as exc_info:
         ray.get(refs)
     assert " Worker unexpectedly exits" not in str(exc_info.value)
 
     # Test a sync case.
     a = RegressionSync.remote()
     a.f.remote()
-    with pytest.raises(
-        (ray.exceptions.RayActorError, ray.exceptions.TaskCancelledError)
-    ) as exc_info:
+    with pytest.raises(ray.exceptions.RayActorError) as exc_info:
         ray.get([a.ping.remote() for _ in range(10000)])
     assert " Worker unexpectedly exits" not in str(exc_info.value)
 

--- a/src/ray/core_worker/task_execution/task_receiver.cc
+++ b/src/ray/core_worker/task_execution/task_receiver.cc
@@ -160,11 +160,20 @@ void TaskReceiver::HandleTask(rpc::PushTaskRequest request,
     };
   };
 
-  auto cancel_callback = [reply](
+  auto cancel_callback = [this, reply](
                              const TaskSpecification &canceled_task_spec,
                              const Status &status,
                              const rpc::SendReplyCallback &canceled_send_reply_callback) {
     if (canceled_task_spec.IsActorTask()) {
+      // For actor tasks, check if cancellation is due to worker shutdown.
+      // This information must travel with the RPC response because the actor death
+      // notification from GCS may arrive at the submitter after this response.
+      {
+        absl::MutexLock lock(&stop_mu_);
+        if (stopping_) {
+          reply->set_worker_exiting(true);
+        }
+      }
       canceled_send_reply_callback(status, nullptr, nullptr);
     } else {
       reply->set_was_cancelled_before_running(true);
@@ -174,16 +183,18 @@ void TaskReceiver::HandleTask(rpc::PushTaskRequest request,
 
   {
     absl::MutexLock lock(&stop_mu_);
+    task_spec = TaskSpecification(std::move(*request.mutable_task_spec()));
     if (stopping_) {
       RAY_LOG(INFO)
           << "Rejecting PushTask due to worker shutdown: task will be cancelled";
       reply->set_was_cancelled_before_running(true);
+      if (task_spec.IsActorTask()) {
+        reply->set_worker_exiting(true);
+      }
       send_reply_callback(
           Status::SchedulingCancelled("Worker is shutting down"), nullptr, nullptr);
       return;
     }
-
-    task_spec = TaskSpecification(std::move(*request.mutable_task_spec()));
 
     if (task_spec.IsActorCreationTask()) {
       SetupActor(task_spec.IsAsyncioActor(),

--- a/src/ray/core_worker/task_execution/task_receiver.cc
+++ b/src/ray/core_worker/task_execution/task_receiver.cc
@@ -165,16 +165,16 @@ void TaskReceiver::HandleTask(rpc::PushTaskRequest request,
                              const Status &status,
                              const rpc::SendReplyCallback &canceled_send_reply_callback) {
     if (canceled_task_spec.IsActorTask()) {
-      // For actor tasks, check if cancellation is due to worker shutdown.
-      // This information must travel with the RPC response because the actor death
-      // notification from GCS may arrive at the submitter after this response.
+      // If task cancelation is due to worker shutdown, propagate that information
+      // to the submitter.
       {
         absl::MutexLock lock(&stop_mu_);
         if (stopping_) {
           reply->set_worker_exiting(true);
+          reply->set_was_cancelled_before_running(true);
         }
       }
-      canceled_send_reply_callback(status, nullptr, nullptr);
+      canceled_send_reply_callback(Status::OK(), nullptr, nullptr);
     } else {
       reply->set_was_cancelled_before_running(true);
       canceled_send_reply_callback(status, nullptr, nullptr);
@@ -185,14 +185,11 @@ void TaskReceiver::HandleTask(rpc::PushTaskRequest request,
     absl::MutexLock lock(&stop_mu_);
     task_spec = TaskSpecification(std::move(*request.mutable_task_spec()));
     if (stopping_) {
-      RAY_LOG(INFO)
-          << "Rejecting PushTask due to worker shutdown: task will be cancelled";
       reply->set_was_cancelled_before_running(true);
       if (task_spec.IsActorTask()) {
         reply->set_worker_exiting(true);
       }
-      send_reply_callback(
-          Status::SchedulingCancelled("Worker is shutting down"), nullptr, nullptr);
+      send_reply_callback(Status::OK(), nullptr, nullptr);
       return;
     }
 

--- a/src/ray/core_worker/task_submission/actor_task_submitter.cc
+++ b/src/ray/core_worker/task_submission/actor_task_submitter.cc
@@ -666,44 +666,9 @@ void ActorTaskSubmitter::HandlePushTaskReply(const Status &status,
     // status.ok() means the worker completed the reply, either succeeded or with a
     // retryable failure (e.g. user exceptions). We complete only on non-retryable case.
     task_manager_.CompletePendingTask(task_id, reply, addr, reply.is_application_error());
-  } else if (status.IsSchedulingCancelled()) {
-    // Check if this is due to actor shutdown or explicit user cancellation.
-    if (reply.worker_exiting()) {
-      // Task cancelled due to actor shutdown - use ACTOR_DIED error
-      rpc::RayErrorInfo error_info;
-      {
-        absl::MutexLock lock(&mu_);
-        auto queue_pair = client_queues_.find(actor_id);
-        if (queue_pair != client_queues_.end() &&
-            queue_pair->second.state_ == rpc::ActorTableData::DEAD) {
-          const auto &death_cause = queue_pair->second.death_cause_;
-          error_info = gcs::GetErrorInfoFromActorDeathCause(death_cause);
-        } else {
-          error_info.set_error_type(rpc::ErrorType::ACTOR_DIED);
-          error_info.set_error_message(
-              "The actor is dead because its worker process has died.");
-        }
-      }
-      RAY_LOG(DEBUG) << "Task " << task_id << " cancelled due to actor " << actor_id
-                     << " shutdown";
-      task_manager_.FailPendingTask(task_spec.TaskId(),
-                                    error_info.error_type(),
-                                    /*status*/ nullptr,
-                                    &error_info);
-    } else {
-      std::ostringstream stream;
-      stream << "The task " << task_id << " is canceled from an actor " << actor_id
-             << " before it executes.";
-      const auto &msg = stream.str();
-      RAY_LOG(DEBUG) << msg;
-      rpc::RayErrorInfo error_info;
-      error_info.set_error_message(msg);
-      error_info.set_error_type(rpc::ErrorType::TASK_CANCELLED);
-      task_manager_.FailPendingTask(task_spec.TaskId(),
-                                    rpc::ErrorType::TASK_CANCELLED,
-                                    /*status*/ nullptr,
-                                    &error_info);
-    }
+  } else if ((status.ok() && reply.was_cancelled_before_running()) ||
+             status.IsSchedulingCancelled()) {
+    HandleTaskCancelledBeforeExecution(status, reply, task_spec);
   } else {
     bool is_actor_dead = false;
     bool fail_immediately = false;
@@ -802,6 +767,88 @@ void ActorTaskSubmitter::HandlePushTaskReply(const Status &status,
     RAY_CHECK(queue_pair != client_queues_.end());
     auto &queue = queue_pair->second;
     queue.cur_pending_calls_--;
+  }
+}
+
+void ActorTaskSubmitter::HandleTaskCancelledBeforeExecution(
+    const Status &status,
+    const rpc::PushTaskReply &reply,
+    const TaskSpecification &task_spec) {
+  const auto task_id = task_spec.TaskId();
+  const auto actor_id = task_spec.ActorId();
+
+  if (reply.worker_exiting()) {
+    // Task cancelled due to actor shutdown - use ACTOR_DIED error.
+    // If we have the death cause, use it immediately. Otherwise,
+    // wait for it from GCS to provide an accurate error message.
+    bool is_actor_dead = false;
+    rpc::RayErrorInfo error_info;
+    {
+      absl::MutexLock lock(&mu_);
+      auto queue_pair = client_queues_.find(actor_id);
+      if (queue_pair != client_queues_.end()) {
+        is_actor_dead = queue_pair->second.state_ == rpc::ActorTableData::DEAD;
+        if (is_actor_dead) {
+          const auto &death_cause = queue_pair->second.death_cause_;
+          error_info = gcs::GetErrorInfoFromActorDeathCause(death_cause);
+        }
+      }
+    }
+
+    if (is_actor_dead) {
+      CancelDependencyResolution(task_id);
+      RAY_LOG(DEBUG) << "Task " << task_id << " cancelled due to actor " << actor_id
+                     << " death";
+      task_manager_.FailPendingTask(task_spec.TaskId(),
+                                    error_info.error_type(),
+                                    /*status*/ nullptr,
+                                    &error_info);
+    } else if (RayConfig::instance().timeout_ms_task_wait_for_death_info() != 0) {
+      CancelDependencyResolution(task_id);
+
+      int64_t death_info_grace_period_ms =
+          current_time_ms() + RayConfig::instance().timeout_ms_task_wait_for_death_info();
+
+      error_info.set_error_type(rpc::ErrorType::ACTOR_DIED);
+      error_info.set_error_message(
+          "The actor is dead because its worker process has died.");
+
+      {
+        absl::MutexLock lock(&mu_);
+        auto queue_pair = client_queues_.find(actor_id);
+        RAY_CHECK(queue_pair != client_queues_.end());
+        auto &queue = queue_pair->second;
+        queue.wait_for_death_info_tasks_.push_back(
+            std::make_shared<PendingTaskWaitingForDeathInfo>(
+                death_info_grace_period_ms, task_spec, status, error_info));
+        RAY_LOG(INFO).WithField(task_spec.TaskId())
+            << "Task cancelled during actor shutdown, waiting for death info from GCS"
+            << ", wait_queue_size=" << queue.wait_for_death_info_tasks_.size();
+      }
+    } else {
+      CancelDependencyResolution(task_id);
+      error_info.set_error_type(rpc::ErrorType::ACTOR_DIED);
+      error_info.set_error_message(
+          "The actor is dead because its worker process has died.");
+      task_manager_.FailPendingTask(task_spec.TaskId(),
+                                    rpc::ErrorType::ACTOR_DIED,
+                                    /*status*/ nullptr,
+                                    &error_info);
+    }
+  } else {
+    // Explicit user cancellation - use TASK_CANCELLED error.
+    std::ostringstream stream;
+    stream << "The task " << task_id << " is canceled from an actor " << actor_id
+           << " before it executes.";
+    const auto &msg = stream.str();
+    RAY_LOG(DEBUG) << msg;
+    rpc::RayErrorInfo error_info;
+    error_info.set_error_message(msg);
+    error_info.set_error_type(rpc::ErrorType::TASK_CANCELLED);
+    task_manager_.FailPendingTask(task_spec.TaskId(),
+                                  rpc::ErrorType::TASK_CANCELLED,
+                                  /*status*/ nullptr,
+                                  &error_info);
   }
 }
 

--- a/src/ray/core_worker/task_submission/actor_task_submitter.cc
+++ b/src/ray/core_worker/task_submission/actor_task_submitter.cc
@@ -667,18 +667,42 @@ void ActorTaskSubmitter::HandlePushTaskReply(const Status &status,
     // retryable failure (e.g. user exceptions). We complete only on non-retryable case.
     task_manager_.CompletePendingTask(task_id, reply, addr, reply.is_application_error());
   } else if (status.IsSchedulingCancelled()) {
-    std::ostringstream stream;
-    stream << "The task " << task_id << " is canceled from an actor " << actor_id
-           << " before it executes.";
-    const auto &msg = stream.str();
-    RAY_LOG(DEBUG) << msg;
+    // Check if this is due to actor shutdown or explicit user cancellation.
+    // Actor shutdown should raise RayActorError, not TaskCancelledError.
+    bool is_actor_dead = false;
     rpc::RayErrorInfo error_info;
-    error_info.set_error_message(msg);
-    error_info.set_error_type(rpc::ErrorType::TASK_CANCELLED);
-    task_manager_.FailPendingTask(task_spec.TaskId(),
-                                  rpc::ErrorType::TASK_CANCELLED,
-                                  /*status*/ nullptr,
-                                  &error_info);
+    {
+      absl::MutexLock lock(&mu_);
+      auto queue_pair = client_queues_.find(actor_id);
+      if (queue_pair != client_queues_.end()) {
+        is_actor_dead = queue_pair->second.state_ == rpc::ActorTableData::DEAD;
+        if (is_actor_dead) {
+          const auto &death_cause = queue_pair->second.death_cause_;
+          error_info = gcs::GetErrorInfoFromActorDeathCause(death_cause);
+        }
+      }
+    }
+
+    if (is_actor_dead) {
+      RAY_LOG(DEBUG) << "Task " << task_id << " cancelled due to actor " << actor_id
+                     << " death";
+      task_manager_.FailPendingTask(task_spec.TaskId(),
+                                    error_info.error_type(),
+                                    /*status*/ nullptr,
+                                    &error_info);
+    } else {
+      std::ostringstream stream;
+      stream << "The task " << task_id << " is canceled from an actor " << actor_id
+             << " before it executes.";
+      const auto &msg = stream.str();
+      RAY_LOG(DEBUG) << msg;
+      error_info.set_error_message(msg);
+      error_info.set_error_type(rpc::ErrorType::TASK_CANCELLED);
+      task_manager_.FailPendingTask(task_spec.TaskId(),
+                                    rpc::ErrorType::TASK_CANCELLED,
+                                    /*status*/ nullptr,
+                                    &error_info);
+    }
   } else {
     bool is_actor_dead = false;
     bool fail_immediately = false;

--- a/src/ray/core_worker/task_submission/actor_task_submitter.cc
+++ b/src/ray/core_worker/task_submission/actor_task_submitter.cc
@@ -668,24 +668,24 @@ void ActorTaskSubmitter::HandlePushTaskReply(const Status &status,
     task_manager_.CompletePendingTask(task_id, reply, addr, reply.is_application_error());
   } else if (status.IsSchedulingCancelled()) {
     // Check if this is due to actor shutdown or explicit user cancellation.
-    // Actor shutdown should raise RayActorError, not TaskCancelledError.
-    bool is_actor_dead = false;
-    rpc::RayErrorInfo error_info;
-    {
-      absl::MutexLock lock(&mu_);
-      auto queue_pair = client_queues_.find(actor_id);
-      if (queue_pair != client_queues_.end()) {
-        is_actor_dead = queue_pair->second.state_ == rpc::ActorTableData::DEAD;
-        if (is_actor_dead) {
+    if (reply.worker_exiting()) {
+      // Task cancelled due to actor shutdown - use ACTOR_DIED error
+      rpc::RayErrorInfo error_info;
+      {
+        absl::MutexLock lock(&mu_);
+        auto queue_pair = client_queues_.find(actor_id);
+        if (queue_pair != client_queues_.end() &&
+            queue_pair->second.state_ == rpc::ActorTableData::DEAD) {
           const auto &death_cause = queue_pair->second.death_cause_;
           error_info = gcs::GetErrorInfoFromActorDeathCause(death_cause);
+        } else {
+          error_info.set_error_type(rpc::ErrorType::ACTOR_DIED);
+          error_info.set_error_message(
+              "The actor is dead because its worker process has died.");
         }
       }
-    }
-
-    if (is_actor_dead) {
       RAY_LOG(DEBUG) << "Task " << task_id << " cancelled due to actor " << actor_id
-                     << " death";
+                     << " shutdown";
       task_manager_.FailPendingTask(task_spec.TaskId(),
                                     error_info.error_type(),
                                     /*status*/ nullptr,
@@ -696,6 +696,7 @@ void ActorTaskSubmitter::HandlePushTaskReply(const Status &status,
              << " before it executes.";
       const auto &msg = stream.str();
       RAY_LOG(DEBUG) << msg;
+      rpc::RayErrorInfo error_info;
       error_info.set_error_message(msg);
       error_info.set_error_type(rpc::ErrorType::TASK_CANCELLED);
       task_manager_.FailPendingTask(task_spec.TaskId(),

--- a/src/ray/core_worker/task_submission/actor_task_submitter.h
+++ b/src/ray/core_worker/task_submission/actor_task_submitter.h
@@ -256,6 +256,18 @@ class ActorTaskSubmitter : public ActorTaskSubmitterInterface {
           timeout_error_info_(std::move(timeout_error_info)) {}
   };
 
+  /// Handle a task that was cancelled before it could execute.
+  /// This method determines whether the cancellation was due to:
+  /// 1. Actor shutdown (worker exiting): If so, raise RayActorError.
+  /// 2. Explicit user cancellation: If so, raise TaskCancelledError.
+  ///
+  /// \param status The RPC status from PushTask.
+  /// \param reply The PushTaskReply message containing cancellation details.
+  /// \param task_spec The specification of the task that was cancelled.
+  void HandleTaskCancelledBeforeExecution(const Status &status,
+                                          const rpc::PushTaskReply &reply,
+                                          const TaskSpecification &task_spec);
+
   struct ClientQueue {
     ClientQueue(bool allow_out_of_order_execution,
                 int32_t max_pending_calls,

